DRILL-6997: Semijoin is changing the join ordering for some tpcds que…

HanumathRao opened the pull request:
…ries.

After transforming DrillJoin -- (DrillRel, DrillAgg) to DrillSemiJoinRel, join ordering is not considering the Semijoin with other joins in the query. This is resulting in sub-optimal plans during the planning phase. The following code changes are made to resolve the issue.
a) A new rule DrillSemiJoinRule is introduced to transform the above pattern to DrillSemiJoinRel.
b) RowType of the DrillSemiJoinRel is same as that of Drill's join relations. 
c) validity checks are overloaded to check for the appropriate row type.

@amansinha100  Can you please review this PR.

amansinha100 commented on exec/java-exec/src/main/java/org/apache/drill/exec/planner/sql/handlers/DefaultSqlHandler.java:
```diff
@@ -277,6 +277,10 @@ protected DrillRel convertToRawDrel(final RelNode relNode) throws SqlUnsupported
         }
       }
 
+      if (context.getPlannerSettings().isSemiJoinEnabled() &&
```
Pls add a brief explanation as comments for why the semi join conversion is done at this stage of the planning process after the normal join ordering is done.

amansinha100 commented on exec/java-exec/src/main/java/org/apache/drill/exec/planner/logical/DrillSemiJoinRule.java:
```diff
@@ -0,0 +1,182 @@
+   * from its input rowtype then do not convert the join to a semi join.
+   */
+  private static boolean isRowTypeSame(RelNode join, RelNode left, RelNode rightInput) {
+    return join.getRowType().getFieldCount() == left.getRowType().getFieldCount() +
```
I am wondering if this should check for field count == 1 or leave it general ? If we leave it general, then should make sure to test correctness. My preference is to restrict it.

amansinha100 commented on exec/java-exec/src/main/java/org/apache/drill/exec/planner/logical/DrillSemiJoinRule.java:
```diff
@@ -0,0 +1,182 @@
+import java.util.function.Predicate;
+
+/**
+ * Planner rule that creates a {@code DrillSemiJoinRule} from a
```
creates a DrillSemiJoinRel (not Rule)

amansinha100 commented on exec/java-exec/src/main/java/org/apache/drill/exec/planner/logical/DrillSemiJoinRel.java:
```diff
@@ -99,6 +104,49 @@ public LogicalOperator implement(DrillImplementor implementor) {
+  @Override public RelDataType deriveRowType() {
+    return SqlValidatorUtil.deriveJoinRowType(
+            left.getRowType(),
+            right.getRowType(),
```
The right side's row type being projected is a key difference compared to the Calcite's row type, so pls add an explicit comment here for it.

HanumathRao commented:
@amansinha100  Thank you for the review. I have addressed the review comments. Regarding the multi column predicate testing for semi-join this test case is already added in the previous checkin. It produces the semi-join plan. I have not checked for the correctness of the result in the test case because this is having an issue when assertions are enabled in the runtime code. I will talk to @Ben-Zvi  and open an issue. However, I have checked for the correctness manually and it does produce correct rows.

amansinha100 commented on exec/java-exec/src/main/java/org/apache/drill/exec/planner/sql/handlers/DefaultSqlHandler.java:
```diff
@@ -277,6 +277,12 @@ protected DrillRel convertToRawDrel(final RelNode relNode) throws SqlUnsupported
       }
 
+      /*Ideally this conversion can be handled during logical planning phase itself but currently the join ordering algorithm in calcite
+        is not considering the semi-joins. This can lead to sub optimal plans. Hence converting the joins to semi-joins post join planning.*/
```
For future reference, pls add the CALCITE-xxx JIRA number for this.

HanumathRao commented:
@amansinha100  Thank you for the review. I have addressed the comments.
